File: LeetCode/LeetCode 20. 有效的括号.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map <char, int> ss;
        ss['('] = 1;
        ss[')'] = -1;
        ss['['] = 2;
        ss[']'] = -2;
        ss['{'] = 3;
        ss['}'] = -3;
        stack <int> pp;
        for(auto x : s) {
            if(ss[x] > 0) {
                pp.push(ss[x]);
            }
            else {
                if(!pp.empty() && pp.top() * -1 == ss[x]) pp.pop();
                else return false;
            }
        }
        return pp.empty();
    }
};

File: LeetCode/LeetCode 20. 有效的括号.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) & 1:
            return False
            
        dic = {")": "(", "]": "[", "}": "{"}
        ss = list()

        for x in s:
            if x in dic:
                if not ss or ss[-1] != dic[x]:
                    return False
                ss.pop()
            else:
                ss.append(x)

        return not ss